_posts/2023-07-24-native-startup-rss-troubleshooting.adoc

```diff
@@ -0,0 +1,230 @@
+we were able to narrow down the issue to the shape of the flamegraphs for the `mmap` system calls.
+To obtain the flamegraphs, we capture the `sys_enter_mmap` system call events with `perf record`,
+and then we generate flamegraphs with the recorded data:
+
+[source,bash]
+----
+$ sudo perf record -e syscalls:sys_enter_mmap --call-graph dwarf -- getting-started/target/getting-started-1.0.0-SNAPSHOT-runner -Xmx64m
+...
+2023-07-21 10:13:11,304 INFO  [io.quarkus] (main) getting-started 1.0.0-SNAPSHOT native (powered by Quarkus 2.7.6.Final) started in 0.023s. Listening on: http://0.0.0.0:8080
+
+$ perf script -i perf.data > out.stacks
+$ /opt/FlameGraph/stackcollapse-perf.pl < out.stacks | /opt/FlameGraph/flamegraph.pl \
+   --color=mem \
+   --title="mmap Flame Graph" \
+   --countname="calls" > out.svg
+----
+
+[NOTE]
+====
+The native executable above was created passing in
+`-Dquarkus.native.debug.enabled` and `-Dquarkus.native.additional-build-args=-H:-DeleteLocalSymbols`
+as extra build flags.
+====
+
+The areas of interest in the flamegraphs are those stacktraces that contain `ApplicationImpl.doStart()` method call.
+If you focus the flamegraph on that method,
+and then click on `Search` at the top right and type `allocate`,
+you can observe which of the `sys_enter_mmap` system calls are related to native heap memory allocation.
+These allocations are for heap chunks, each of which is 1MB,
+which are part of the thread local allocation logic in GraalVM’s runtime engine, SubstrateVM
+(further details in the
+https://quarkus.io/guides/native-reference#rss[Quarkus native reference guide]
+).
+
+For Quarkus 2.7.6.Final, the flamegraph looks like this:
+
+image::rss-before-mmap.svg[startup rss startup flamegraph for Quarkus 2.7.6.Final]
+
+And here's a screenshot of those mmap calls that are related heap chunk allocations:
+
+image::rss-before-mmap-allocate.png[heap chunk allocation related mmap calls for Quarkus 2.7.6.Final]
+
+If we count the number of times `allocate` is highlighted above, the number is 2.
+So, we can roughly say that the `main` thread in Quarkus 2.7.6.Final will allocate 2MB of heap chunks.
+
+For Quarkus 2.13.4.Final, the flamegraph looks like this:
+
+image::rss-after-mmap.svg[startup rss startup flamegraph for Quarkus 2.13.4.Final]
+
+If we focus on the same area, we observe the count is 5, so that would be roughly 5MB of allocated heap chunks for the `main` thread:
+
+image::rss-after-mmap-allocate.png[heap chunk allocation related mmap calls for Quarkus 2.13.4.Final]
+
+The increase in object allocation between these two Quarkus versions at runtime startup,
+combined with the idiosyncrasies of how thread local allocation works in SubstrateVM,
+is what caused this increase in RSS consumption.
+
+Since then, improvements were made to the components that caused the extra object allocation.
```

Comment:
Right, I left that information out.

The idea is to look at the components that appear in the stacktrace leading to the slow allocation and see what changes have happened in those. This is not a perfect science since it could happen that a slow allocation results from somewhere that is not responsible for the increase (by luck), but it gives some good hints. In this case it was changes in how the SmallRye configuration was handled at runtime.

Shall I add something along those lines?